generator/bench/Bench/Pos/Criterion/Block/Logic.hs

```diff
@@ -0,0 +1,243 @@
+import           Pos.Core.Configuration (genesisBlockVersionData, genesisData,
+                     genesisSecretKeys, slotSecurityParam)
+import           Pos.Core.Genesis (FakeAvvmOptions (..), GenesisData (..),
+                     GenesisInitializer (..), TestnetBalanceOptions (..))
+import           Pos.Core.Slotting (Timestamp (..), epochIndexL, getEpochOrSlot)
+import           Pos.Core.Update (BlockVersionData (..))
+import           Pos.Crypto.Configuration (ProtocolMagic (..))
+import           Pos.DB (getTipHeader)
+import           Pos.DB.DB (initNodeDBs)
+import           Pos.Generator.Block (BlockGenParams (..), TxGenParams (..),
+                     genBlockNoApply, genBlocks, mkBlockGenContext)
+import           Pos.Launcher.Configuration (ConfigurationOptions (..),
+                     HasConfigurations, defaultConfigurationOptions,
+                     withConfigurationsM)
+import           Pos.Lrc.Context (lrcActionOnEpochReason)
+import qualified Pos.Lrc.DB as LrcDB
+import           Pos.Txp.Logic.Global (txpGlobalSettings)
+import           Pos.Util.CompileInfo (withCompileInfo)
+import           Pos.Util.Util (realTime)
+
+import           Test.Pos.Block.Logic.Emulation (runEmulation, sudoLiftIO)
+import           Test.Pos.Block.Logic.Mode (BlockTestContext, BlockTestMode,
+                     TestParams (..), initBlockTestContext, runBlockTestMode)
+import           Test.Pos.Block.Logic.Util (satisfySlotCheck)
+
+-- | Criterion configuration
+config :: Config
+config = defaultConfig
+    { reportFile = Just "verification.html"
+    , resamples = 10
+    , timeLimit = 5.0
+    , verbosity = Verbose
+    }
+
+genesisInitializer :: GenesisInitializer
+genesisInitializer = GenesisInitializer
+    { giTestBalance = balance
+    , giFakeAvvmBalance = FakeAvvmOptions
+          { faoCount = 1
+          , faoOneBalance = maxBound
+          }
+    , giAvvmBalanceFactor = unsafeCoinPortionFromDouble 0
+    , giUseHeavyDlg = False
+    , giSeed = 0
+    }
+
+balance :: TestnetBalanceOptions
+balance = TestnetBalanceOptions
+    { tboPoors = 1
+    , tboRichmen = 1
+    , tboTotalBalance = maxBound
+    , tboRichmenShare = 1
+    , tboUseHDAddresses = False
+    }
+
+runBTM
+    :: TestParams
+    -> BlockTestContext
+    -> BlockTestMode a
+    -> IO a
+runBTM tp ctx btm = runEmulation (getTimestamp (_tpStartTime tp)) $ runReaderT btm ctx
+
+-- | Benchmark which runs `verifyAndApplyBlocks` && `rollbackBlocks`.
+verifyBlocksBenchmark
+    :: HasConfigurations
+    => ProtocolMagic
+    -> TestParams
+    -> BlockTestContext
+    -> Benchmark
+verifyBlocksBenchmark !pm !tp !ctx =
+    bgroup "block verification"
+        [ env (runBlockTestMode tp (genEnv (BlockCount 100)))
+            $ \ ~(vctx, blocks) -> bench "verifyAndApplyBlocks" (verifyAndApplyBlocksB vctx blocks)
+        -- `verifyBlocksPrefix` will succeed only on the first block, it
+        -- requires that blocks are applied.
```

Comment:
Wha? Then what's the use of it?

Reply:
That's a stale comment, I'll remove it.  At a time I tried to generate blocks without applying them, and then only the first block was successfully verified and the verification was failing on the tail.